### PR TITLE
[WIP] Upgraded to Preact 10.3.4

### PR DESCRIPTION
**Pull request by nickytonline**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Preact has been upgraded to the latest version, Preact X. This gives us hooks, componentDidCatch, fragments etc. See the [official release](https://preactjs.com/guide/v10/whats-new/) for the whole rundown. Here is the [official migration guide](https://preactjs.com/guide/v10/upgrade-guide) for Preact 8.x to X if you're curious why some changes have been made.

I've done a smoke test of the site on my local. Onboarding worked fine, search is good and the site seems to behaving normally.

Storybook appears to be broken still. This is related to work that needs to be done in #3490 and #2034. I will address that in a separate PR.

## Related Tickets & Documents

#4243 

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

Some cool X gif

### CODE REVIEW

File: app/javascript/article-form/articleForm.jsx [hunk 1]
```diff
@@ -1,4 +1,3 @@
-import 'preact/devtools';
```
Comment (nickytonline) [inline app/javascript/article-form/articleForm.jsx]: I've removed `preact/devtools` as it no longer exists. `preact/debug` is what should be used, but it should only be for development purposes. I will address this in another PR.

See [Debugging Preact Apps](https://preactjs.com/guide/v10/debugging/#strip-devtools-from-production)

File: app/javascript/packs/Search.jsx [hunk 2]
```diff
@@ -1,9 +1,9 @@
 import { h, render } from 'preact';
 import { Search } from '../src/components/Search';
-import 'focus-visible'
+import 'focus-visible';
```
Comment (nickytonline) [inline app/javascript/packs/Search.jsx]: A little prettier action happening in a few files.

File: app/javascript/packs/Search.jsx [hunk 3]
```diff
@@ -1,9 +1,9 @@
...
 
 document.addEventListener('DOMContentLoaded', () => {
   const root = document.getElementById('nav-search-form-root');
 
-  render(<Search />, root, root.firstElementChild);
+  render(<Search />, root);
```
Comment (nickytonline) [inline app/javascript/packs/Search.jsx]: This is part of the upgrade. VDOM diffing works the way it does in most other libraries, e.g. React.

**Comment by nickytonline** [comment]
The codeclimate issues appear to be complaining about code duplication for `PropTypes` but it's not an issue.  it can be ignored. The prop types are different in each case.

**Comment by nickytonline** [comment]
I need to sort out the tests in the frontend as a lot of the snapshots changed, but the structure is different and information appears to be missing. Maybe Preact X is not compatible with some of the tools we were using for snapshot testing with Preact. Either way, gotta get it sorted. 😉

image

**Comment by nickytonline** [comment]
Tiny update. It looks like there are issues with `preact-render-spy` and Preact X. See https://github.com/mzgoddard/preact-render-spy/issues/92.

File: app/javascript/article-form/articleForm.jsx (hunk 1 above)
Comment (maestromac) [reply app/javascript/article-form/articleForm.jsx]: Thanks for catching this!

**Comment by nickytonline** [comment]
I spoke briefly with Jason Miller from Preact, and he is going to look into getting preact-render-spy etc. fixed up. So for the time being, I will put this PR on hold.

**Comment by joshpuetz** [comment]
@nickytonline Seems like https://github.com/mzgoddard/preact-render-spy might be abandoned: there's a Preact X fork up at https://github.com/shelacek/preact-render-spy. What do we use `preact-render-spy` for anyway? Is using a pure enzyme approach like https://github.com/preactjs/enzyme-adapter-preact-pure a replacement?

**Comment by nickytonline** [comment]
> @nickytonline Seems like [mzgoddard/preact-render-spy](https://github.com/mzgoddard/preact-render-spy) might be abandoned: there's a Preact X fork up at [shelacek/preact-render-spy](https://github.com/shelacek/preact-render-spy). What do we use `preact-render-spy` for anyway? Is using a pure enzyme approach like [preactjs/enzyme-adapter-preact-pure](https://github.com/preactjs/enzyme-adapter-preact-pure) a replacement?

@joshpuetz, I tried the fork of preact-render-spy and it does not work. We use preact-render-spy for snapshot testing. It does appear to be abandoned although Jason Miller talked about fixing it up. See this [Tweet thread](https://twitter.com/nickytonline/status/1019413590451908613) between Jason and I.

Also, I'm looking to move away from snapshot testing in favour of preact-testing-library. Even enzyme is not considered to be the way to test React components now (so Preact as well). As I mentioned on Slack, [react-testing-library is the recommended tool](https://reactjs.org/docs/testing.html#tools) by the React team, so I would be inclined to move towards preact-testing-library.

The way I see it panning out is migrating to another testing tool and once all tests are migrated, we can upgrade.

**Comment by nickytonline** [comment]
Closing this @Ridhwana as we'll need to make a new PR which will include the new testing tools.

**Pull request closed without merging**